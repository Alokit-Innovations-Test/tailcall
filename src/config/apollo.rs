use anyhow::Context;
use serde::{Deserialize, Serialize};

use crate::config::ConfigReaderContext;
use crate::mustache::Mustache;

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Apollo {
    ///
    /// Setting `apiKey` for Apollo.
    pub api_key: String,
    ///
    /// Setting `graphRef` for Apollo in the format <graphId>@<variant>.
    pub graph_ref: String,
    ///
    /// Setting `userVersion` for Apollo.
    #[serde(default = "default_user_version")]
    pub user_version: String,
    ///
    /// Setting `platform` for Apollo.
    #[serde(default = "default_platform")]
    pub platform: String,
    ///
    /// Setting `version` for Apollo.
    #[serde(default = "default_version")]
    pub version: String,
}

fn default_user_version() -> String {
    "1.0".to_string()
}

fn default_platform() -> String {
    "platform".to_string()
}

fn default_version() -> String {
    "1.0".to_string()
}

impl Apollo {
    pub fn render_mustache(&mut self, reader_ctx: &ConfigReaderContext) -> anyhow::Result<()> {
        let Apollo { api_key, graph_ref, user_version, platform, version } = self;

        let api_key_tmpl = Mustache::parse(api_key)?;
        *api_key = api_key_tmpl
            .render_string(reader_ctx)
            .context("apiKey is not defined")?;

        let graph_ref_tmpl = Mustache::parse(graph_ref)?;
        *graph_ref = graph_ref_tmpl
            .render_string(reader_ctx)
            .context("graphRef is not defined")?;

        let user_version_tmpl = Mustache::parse(user_version)?;
        if let Some(rendered) = user_version_tmpl.render_string(reader_ctx) {
            *user_version = rendered;
        }

        let platform_tmpl = Mustache::parse(platform)?;
        if let Some(rendered) = platform_tmpl.render_string(reader_ctx) {
            *platform = rendered;
        }

        let version_tmpl = Mustache::parse(version)?;
        if let Some(rendered) = version_tmpl.render_string(reader_ctx) {
            *version = rendered;
        }

        Ok(())
    }
}
