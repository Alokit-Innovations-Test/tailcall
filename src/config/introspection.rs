use std::collections::BTreeMap;

use async_graphql::futures_util::future::join_all;
use reqwest;
use serde_json::json;

// GraphQL introspection response types.
#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionResult {
  pub data: Option<IntrospectionData>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionData {
  pub __schema: IntrospectionSchema,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionSchema {
  #[serde(rename = "queryType")]
  pub query_type: IntrospectionType,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionType {
  pub fields: Vec<Field>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Field {
  pub name: String,
  pub args: Option<Vec<Arg>>,
  #[serde(rename = "type")]
  pub type_: Option<Type_>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Arg {
  pub name: String,
  #[serde(rename = "type")]
  pub type_: Type_,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Type_ {
  pub name: Option<String>,
  pub kind: Option<String>,
  #[serde(rename = "ofType")]
  pub of_type: Option<Box<Type_>>,
  pub fields: Option<Vec<Field>>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum TypeKind {
  Scalar,
  Object,
  Interface,
  Union,
  Enum,
  InputObject,
  List,
  NonNull,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
pub struct IntrospectionResults(pub BTreeMap<String, IntrospectionResult>);

impl IntrospectionResults {
  pub fn merge_right(mut self, other: Self) -> Self {
    let mut merged = self.0;
    for (key, value) in other.0 {
      merged.insert(key, value);
    }
    self.0 = merged;
    self
  }
}

pub async fn introspect_endpoint(graphql_url: &String) -> Result<IntrospectionResult, Box<dyn std::error::Error>> {
  let introspection_query: String =
  json!({"query": "query { __schema { queryType { name fields { name args { name type { name kind ofType {name }}} type { name kind ofType { name fields {name} } fields { name }}  } } } }"}).to_string();

  let result = reqwest::Client::new()
    .post(graphql_url)
    .header("Content-Type", "application/json")
    .body(introspection_query)
    .send()
    .await?
    .json::<IntrospectionResult>()
    .await?;

  Ok(result)
}

pub async fn introspect_endpoints(graphql_urls: Vec<String>) -> IntrospectionResults {
  let mut results = BTreeMap::new();
  let introspect_futures = graphql_urls
    .iter()
    .map(|url| async move { (url.clone(), introspect_endpoint(url).await.unwrap()) });
  let joined = join_all(introspect_futures).await;
  for (url, introspection_result) in joined {
    results.insert(url.clone(), introspection_result.clone());
  }
  IntrospectionResults(results)
}

pub fn get_arg_type(
  introspection_result: &IntrospectionResult,
  query_name: &String,
  arg_name: &String,
) -> Option<String> {
  introspection_result
    .data
    .as_ref()
    .and_then(|data| {
      data
        .__schema
        .query_type
        .fields
        .iter()
        .find(|field| field.name == *query_name)
    })
    .and_then(|field| field.args.as_ref())
    .and_then(|args| args.iter().find(|arg| arg.name == *arg_name))
    .and_then(|arg| {
      arg.type_.name.as_ref().map_or_else(
        || {
          let kind = arg.type_.kind.clone().unwrap_or_default();
          match &arg.type_.of_type {
            Some(type_) => type_.name.as_ref().map(|name| format_name(name, &kind)),
            None => None,
          }
        },
        |name| Some(name.clone()),
      )
    })
}

fn format_name(name: &String, kind: &String) -> String {
  if kind == "LIST" {
    format!("[{}]", name)
  } else if kind == "NON_NULL" {
    format!("{}!", name)
  } else {
    name.clone()
  }
}

#[cfg(test)]
mod tests {
  use std::fs;
  use std::path::PathBuf;

  use httpmock::Method::POST;
  use httpmock::MockServer;

  use crate::config::introspection::*;

  fn load_mock_introspection_result() -> String {
    let mut mock_result_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    mock_result_path.push("tests/data/introspection-result.json");
    fs::read_to_string(mock_result_path).unwrap()
  }

  fn mock_introspection_results() -> BTreeMap<String, IntrospectionResult> {
    let result: IntrospectionResult = serde_json::from_str(&load_mock_introspection_result()).unwrap();
    let mut introspection_result: BTreeMap<String, IntrospectionResult> = BTreeMap::new();
    introspection_result.insert("http://localhost:8000/graphql".to_string(), result);
    introspection_result
  }

  #[test]
  fn test_get_arg_type() {
    let introspection_results = mock_introspection_results();
    let introspection_result = introspection_results
      .get(&"http://localhost:8000/graphql".to_string())
      .unwrap();
    assert_eq!(
      get_arg_type(introspection_result, &"user".to_string(), &"id".to_string()).unwrap(),
      "Int"
    );
    assert_eq!(
      get_arg_type(introspection_result, &"post".to_string(), &"id".to_string()).unwrap(),
      "Int!"
    );
    assert_eq!(
      get_arg_type(introspection_result, &"post".to_string(), &"doesntexist".to_string()),
      None
    );
    assert_eq!(
      get_arg_type(introspection_result, &"doesntexist".to_string(), &"id".to_string()),
      None
    );
  }

  #[tokio::test]
  async fn test_introspect_endpoint() {
    let server = MockServer::start();
    let result = load_mock_introspection_result();

    server.mock(|when, then| {
      when.method(POST).path("/graphql");
      then.status(200).header("content-type", "application/json").body(result);
    });

    let graphql_url = server.url("/graphql");

    let introspect_result = introspect_endpoint(&graphql_url).await;
    assert_eq!(
      introspect_result
        .unwrap()
        .data
        .unwrap()
        .__schema
        .query_type
        .fields
        .len(),
      2
    );
  }
}
