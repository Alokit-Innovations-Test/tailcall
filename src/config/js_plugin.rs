use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
pub struct JsPlugin {
  pub src: String,
}

impl JsPlugin {
  pub fn merge_right(self, other: Self) -> Self {
    Self {
      src: if self.src.is_empty() {
        other.src
      } else {
        self.src
      }
    }
  }
}
