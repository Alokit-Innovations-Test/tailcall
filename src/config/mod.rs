mod config;
mod from_document;
pub mod group_by;
mod into_document;
mod key_values;
mod n_plus_one;
mod server;
mod source;
mod js_plugin;

pub use config::*;
pub use key_values::*;
pub use server::*;
pub use source::*;
pub use js_plugin::*;

fn is_default<T: Default + Eq>(val: &T) -> bool {
  *val == T::default()
}
