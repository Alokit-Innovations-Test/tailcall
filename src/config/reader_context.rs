use std::collections::BTreeMap;
use std::sync::Arc;

use crate::path_value::PathValue;
use crate::EnvIO;

pub struct ConfigReaderContext<'a> {
    pub env: Arc<dyn EnvIO>,
    pub vars: &'a BTreeMap<String, String>,
}

impl<'a> PathValue for ConfigReaderContext<'a> {
    fn get_path_value<Path>(&self, path: &[Path]) -> Option<async_graphql::Value>
    where
        Path: AsRef<str>,
    {
        path.split_first()
            .and_then(|(head, tail)| match head.as_ref() {
                "vars" => self.vars.get_path_value(tail),
                "env" => self.env.get_path_value(tail),
                _ => None,
            })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::TestEnvIO;

    #[test]
    fn path_value() {
        let reader_context = ConfigReaderContext {
            env: Arc::new(TestEnvIO::from_iter([(
                "ENV_1".to_owned(),
                "ENV_VAL".to_owned(),
            )])),
            vars: &BTreeMap::from_iter([("VAR_1".to_owned(), "VAR_VAL".to_owned())]),
        };

        assert_eq!(
            reader_context.get_path_value(&["env", "ENV_1"]),
            Some("ENV_VAL".into())
        );
        assert_eq!(reader_context.get_path_value(&["env", "ENV_5"]), None);
        assert_eq!(
            reader_context.get_path_value(&["vars", "VAR_1"]),
            Some("VAR_VAL".into())
        );
        assert_eq!(reader_context.get_path_value(&["vars", "VAR_6"]), None);
        assert_eq!(reader_context.get_path_value(&["unknown", "unknown"]), None);
    }
}
