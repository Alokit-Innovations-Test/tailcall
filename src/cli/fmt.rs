use colored::*;

use crate::core::config::{Config, QueryPath};

pub struct Fmt<Out> {
    out: Out,
}

impl<Out: std::io::Write> Fmt<Out> {
    pub fn new(out: Out) -> Self {
        Self { out }
    }
    pub fn heading(&mut self, heading: &str) -> std::io::Result<()> {
        let heading = format!("{}", heading.bold());
        self.append(&heading)
    }

    fn meta(&mut self, meta: &String) -> std::io::Result<()> {
        let val = format!("{}", meta.yellow());
        self.append(&val)
    }

    pub fn append(&mut self, s: &str) -> std::io::Result<()> {
        writeln!(self.out, "{}\n", s)
    }

    /// This function is usually used with BufWriter
    /// It flushes the output and returns the lock
    pub fn display(mut self) -> std::io::Result<Out> {
        self.out.flush()?;
        Ok(self.out)
    }

    /// This function is usually used in replacement of println!()
    /// It flushes the output and drops the lock
    pub fn display_and_drop(mut self) -> std::io::Result<()> {
        self.out.flush()?;
        Ok(())
    }

    fn format_n_plus_one_queries(&mut self, n_plus_one_info: QueryPath) -> std::io::Result<()> {
        self.meta(&n_plus_one_info.to_string())
    }

    pub fn log_n_plus_one(&mut self, show_npo: bool, config: &Config) -> std::io::Result<()> {
        let n_plus_one_info = config.n_plus_one();
        self.append(format!("N + 1 detected: {}", n_plus_one_info.size()).as_str())?;

        if show_npo {
            self.format_n_plus_one_queries(n_plus_one_info)?;
        }
        Ok(())
    }
}
