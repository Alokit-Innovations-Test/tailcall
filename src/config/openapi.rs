use std::collections::BTreeMap;

use oas3::spec::ObjectOrReference;
use oas3::{Schema, Spec};

use crate::config::{Arg, Config, Field, Http, RootSchema, Server, Type, Upstream};

trait ToDebugString: std::fmt::Debug {
    fn to_debug_string(&self) -> String {
        format!("{:?}", self)
    }
}

impl<T: std::fmt::Debug> ToDebugString for T {}

fn map_spec_type(typ: String) -> String {
    if typ.as_str() == "Integer" {
        "Int".to_string()
    } else {
        typ
    }
}

fn get_schema_type<F: FnOnce() -> String>(get_name: F, schema: Schema) -> (bool, String) {
    let typ = schema.schema_type.unwrap().to_debug_string();
    let (is_list, type_name) = match typ.as_str() {
        "Array" => {
            if let ObjectOrReference::Ref { ref ref_path } = schema.items.unwrap().as_ref() {
                (true, ref_path.split('/').last().unwrap().to_string())
            } else {
                unreachable!()
            }
        }
        "Object" => (false, get_name()),
        _ => (false, typ),
    };

    (is_list, map_spec_type(type_name))
}

fn make_config_types(spec: &Spec) -> BTreeMap<String, Type> {
    let components = spec.components.iter().next().cloned().unwrap();
    let mut types = BTreeMap::new();
    let mut query_fields = BTreeMap::new();
    let mut mutation_fields = BTreeMap::new();

    for (mut path, path_item) in spec.paths.clone().into_iter() {
        let (is_query, operation) = [
            path_item.get,
            path_item.head,
            path_item.options,
            path_item.trace,
            path_item.put,
            path_item.post,
            path_item.delete,
            path_item.patch,
        ]
        .into_iter()
        .enumerate()
        .filter_map(|(i, o)| o.map(|o| (i < 4, o)))
        .next()
        .unwrap();
        let output_type = operation
            .responses
            .get("200")
            .unwrap()
            .resolve(spec)
            .unwrap()
            .content
            .get("application/json")
            .cloned()
            .unwrap()
            .schema
            .unwrap();

        let args: BTreeMap<String, Arg> = operation
            .parameters
            .iter()
            .map(|param| {
                let param = param.resolve(spec).unwrap();
                let (is_list, name) =
                    get_schema_type(|| param.param_type.unwrap(), param.schema.unwrap());
                (
                    param.name,
                    Arg {
                        type_of: name,
                        list: is_list,
                        required: param.required.unwrap_or_default(),
                        doc: None,
                        modify: None,
                        default_value: None,
                    },
                )
            })
            .collect();

        let schema = output_type.resolve(spec).unwrap();

        let (is_list, name) = get_schema_type(
            || {
                if let ObjectOrReference::Ref { ref_path } = output_type {
                    ref_path.split('/').last().unwrap().to_string()
                } else {
                    unreachable!()
                }
            },
            schema,
        );

        if !args.is_empty() {
            let re = regex::Regex::new(r"\{\w+\}").unwrap();
            path = re
                .replacen(path.as_str(), 0, |cap: &regex::Captures| {
                    let arg_name = &cap[0][1..cap[0].len() - 1];
                    format!("{{{{args.{}}}}}", arg_name)
                })
                .to_string();
        }

        let field = Field {
            type_of: name,
            list: is_list,
            args,
            http: Some(Http { path, ..Default::default() }),
            ..Default::default()
        };

        let fields = if is_query {
            &mut query_fields
        } else {
            &mut mutation_fields
        };

        fields.insert(operation.operation_id.unwrap(), field);
    }

    if !query_fields.is_empty() {
        types.insert(
            "Query".to_string(),
            Type { fields: query_fields, ..Default::default() },
        );
    }
    if !mutation_fields.is_empty() {
        types.insert(
            "Mutation".to_string(),
            Type { fields: mutation_fields, ..Default::default() },
        );
    }

    for (name, component) in components.schemas.into_iter() {
        let schema = component.resolve(spec).unwrap();
        if schema.schema_type.unwrap().to_debug_string().as_str() == "Array" {
            continue;
        }

        types.insert(
            name,
            Type {
                fields: schema
                    .properties
                    .into_iter()
                    .map(|(name, property)| {
                        (
                            name.clone(),
                            Field {
                                type_of: {
                                    map_spec_type(format!(
                                        "{:?}",
                                        property.resolve(spec).unwrap().schema_type.unwrap()
                                    ))
                                },
                                required: schema.required.contains(&name),
                                ..Default::default()
                            },
                        )
                    })
                    .collect(),
                ..Default::default()
            },
        );
    }

    types
}

pub fn config_from_openapi_spec(content: &str) -> Result<Config, anyhow::Error> {
    let spec = oas3::from_reader(content.as_bytes()).unwrap();
    let types = make_config_types(&spec);
    let config = Config {
        server: Server { graphiql: Some(true), ..Default::default() },
        upstream: Upstream {
            base_url: spec.servers.first().cloned().map(|server| server.url),
            ..Default::default()
        },
        schema: RootSchema {
            query: types.get("Query").map(|_| "Query".into()),
            mutation: types.get("Mutation").map(|_| "Mutation".into()),
            ..Default::default()
        },
        types,
        unions: Default::default(),
        links: vec![],
        telemetry: Default::default(),
    };

    println!("{}", config.to_sdl());
    Ok(config)
}
