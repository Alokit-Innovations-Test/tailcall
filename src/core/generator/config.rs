use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
