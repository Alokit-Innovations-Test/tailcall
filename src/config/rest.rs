use std::cmp::Ordering;
use std::collections::BTreeMap;
use std::fmt::{self, Display, Formatter, Write};
use std::hash::Hasher;
use std::sync::Arc;

use async_graphql::parser::types::{BaseType, Type as GQLType};
use async_graphql_value::ConstValue;
use itertools::{EitherOrBoth, Itertools};
use serde::de::Visitor;
use serde::{de, ser, Deserialize, Deserializer, Serialize, Serializer};
use serde_json::de::StrRead;

use crate::async_graphql_hyper::GraphQLRequest;
use crate::blueprint::OperationQuery;
use crate::http::Method;

#[derive(Clone, Debug, Default)]
pub struct RestApis(pub BTreeMap<Rest, String>);

impl RestApis {
    pub fn create_operations(&self) -> Vec<OperationQuery> {
        self.0
            .iter()
            .map(|(k, v)| {
                let variables = ConstValue::Object(
                    k.path
                        .variables()
                        .map(|var| {
                            (
                                async_graphql::Name::new(var.name.as_ref()),
                                var.default_value(),
                            )
                        })
                        .collect(),
                );
                let variables = async_graphql::Variables::from_value(variables);
                OperationQuery::new_with_variables(v.into(), "".into(), variables)
            })
            .collect()
    }

    pub fn dispatch_path(
        &self,
        method: hyper::Method,
        path: &str,
    ) -> anyhow::Result<GraphQLRequest> {
        let path = format!("\"{path}\"");
        let mut deserializer = serde_json::Deserializer::new(StrRead::new(path.as_str()));
        let path = RestPath::deserialize(&mut deserializer)?;
        let req_rest = Rest { method: method.try_into()?, path };

        let (rest, query) = self
            .0
            .get_key_value(&req_rest)
            .map(|(k, v)| (k.clone(), v))
            .ok_or(anyhow::anyhow!("path not found"))?;
        let vars_json = rest.path.extract_variable_values_from(req_rest.path);

        let req = async_graphql::Request::new(query)
            .variables(async_graphql::Variables::from_json(vars_json));

        Ok(GraphQLRequest(req))
    }
}

impl PartialEq for RestApis {
    fn eq(&self, other: &Self) -> bool {
        self.0.len() == other.0.len()
            && self
                .0
                .iter()
                .all(|(rest, query)| other.0.get(rest).map_or(false, |q| q.eq(query)))
    }
}

impl Eq for RestApis {}

impl RestApis {
    pub fn merge_right(mut self, other: Self) -> Self {
        self.0.extend(other.0);
        self
    }

    pub fn new() -> Self {
        Self(BTreeMap::new())
    }

    pub fn insert(&mut self, rest: Rest, query: impl Into<String>) {
        self.0.insert(rest, query.into());
    }
}

#[derive(
    Clone, Debug, PartialEq, Deserialize, PartialOrd, Ord, Serialize, Eq, schemars::JsonSchema,
)]
/// The @rest operator creates a rest api for the operation it is applied to
#[serde(rename_all = "camelCase")]
pub struct Rest {
    /// Specifies the path for the rest api, relative to the base url.
    pub path: RestPath,
    /// Specifies the HTTP Method for the rest api
    #[serde(default)]
    pub method: Method,
}

#[derive(Clone, Debug, schemars::JsonSchema)]
pub struct RestPath {
    tokens: Vec<Token>,
}

impl PartialEq for RestPath {
    fn eq(&self, other: &Self) -> bool {
        self.tokens
            .iter()
            .zip_longest(other.tokens.iter())
            .all(|either_or_both| match either_or_both {
                EitherOrBoth::Both(Token::Static(a), Token::Static(b)) => a.eq(b),
                EitherOrBoth::Both(Token::Variable { .. }, _) => true,
                EitherOrBoth::Both(_, Token::Variable { .. }) => true,
                _ => false,
            })
    }
}

impl Eq for RestPath {}

impl PartialOrd for RestPath {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for RestPath {
    fn cmp(&self, other: &Self) -> Ordering {
        match self.tokens.len().cmp(&other.tokens.len()) {
            Ordering::Equal => self
                .tokens
                .iter()
                .zip(other.tokens.iter())
                .find_map(|(a, b)| match (a, b) {
                    (Token::Static(a), Token::Static(b)) => Some(a.cmp(b)),
                    _ => None,
                })
                .unwrap_or(Ordering::Equal),
            res => res,
        }
    }
}

impl RestPath {
    pub fn variables_mut(&mut self) -> impl Iterator<Item = &mut Variable> {
        self.tokens.iter_mut().filter_map(|token| match token {
            Token::Variable(var) => Some(var),
            _ => None,
        })
    }

    pub fn variables(&self) -> impl Iterator<Item = &Variable> {
        self.tokens.iter().filter_map(|token| match token {
            Token::Variable(var) => Some(var),
            _ => None,
        })
    }

    pub fn extract_variable_values_from(self, req_path: RestPath) -> serde_json::Value {
        let mut values = serde_json::Map::new();
        for (token, req_token) in self.tokens.into_iter().zip(req_path.tokens.into_iter()) {
            if let (Token::Variable(var), Token::Static(val)) = (token, req_token) {
                values.insert(
                    var.name.to_string(),
                    serde_json::Value::String(val.to_string()),
                );
            }
        }
        serde_json::Value::Object(values)
    }
}

impl<'de> Deserialize<'de> for RestPath {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct TokensVisitor;

        impl<'de> Visitor<'de> for TokensVisitor {
            type Value = RestPath;

            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
                formatter.write_str("a valid path")
            }

            fn visit_str<E>(self, mut v: &str) -> std::result::Result<Self::Value, E>
            where
                E: de::Error,
            {
                if !v.is_empty() && &v[v.len() - 1..] == "/" {
                    v = &v[..v.len() - 1];
                }

                let mut tokens = v.split('/');

                tokens
                    .next()
                    .filter(|token| token.is_empty())
                    .ok_or(E::custom("path should start with \"/\""))?;

                let tokens = tokens
                    .map(|val| {
                        if let Some(var_name) = val.strip_prefix('$') {
                            let var = Variable { name: var_name.into(), typ: None };
                            Token::Variable(var)
                        } else {
                            Token::Static(val.into())
                        }
                    })
                    .collect();
                Ok(RestPath { tokens })
            }
        }

        deserializer.deserialize_str(TokensVisitor)
    }
}

impl Serialize for RestPath {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut result_str = String::new();
        self.tokens
            .iter()
            .map(|token| write!(&mut result_str, "{token}"))
            .collect::<anyhow::Result<Vec<_>, _>>()
            .map_err(|err| <S::Error as ser::Error>::custom(format!("{err}")))?;

        serializer.serialize_str(&result_str)
    }
}

#[derive(Clone, Debug, PartialEq, Hash, Eq, schemars::JsonSchema)]
pub enum Token {
    Static(Arc<str>),
    Variable(Variable),
}

#[derive(Clone, Debug, PartialEq, Eq, schemars::JsonSchema)]
pub struct Variable {
    pub name: Arc<str>,
    #[serde(skip)]
    pub typ: Option<GQLType>,
}

pub fn gql_type_default_value(typ: &GQLType) -> ConstValue {
    match typ {
        GQLType { nullable: true, .. } => ConstValue::Null,
        GQLType { base, .. } => match base {
            BaseType::Named(name) => match name.as_str() {
                "Int" => 0.into(),
                "String" => "".into(),
                "Boolean" => false.into(),
                "Float" => 0.0.into(),
                _ => ConstValue::Null,
            },
            BaseType::List(typ) => ConstValue::List(vec![gql_type_default_value(typ.as_ref())]),
        },
    }
}

impl Variable {
    pub fn default_value(&self) -> ConstValue {
        match &self.typ {
            None => ConstValue::Null,
            Some(typ) => gql_type_default_value(typ),
        }
    }
}

impl std::hash::Hash for Variable {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.name.hash(state)
    }
}

impl PartialOrd for Variable {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.name.cmp(&other.name))
    }
}

impl Ord for Variable {
    fn cmp(&self, other: &Self) -> Ordering {
        self.name.cmp(&other.name)
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let str = match self {
            Token::Static(val) => val.to_string(),
            Token::Variable(var) => format!("${}", var.name),
        };
        write!(f, "{}", str)
    }
}
