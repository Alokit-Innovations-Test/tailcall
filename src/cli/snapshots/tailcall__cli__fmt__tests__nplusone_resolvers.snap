---
source: src/cli/fmt.rs
expression: "formatted.join(\"\\n\")"
---
[33mquery { f1 { f2 } }[0m
